### PR TITLE
Add fast-path to format data

### PR DESCRIPTION
**Pull request by coroa**
Co-authored-by: Matthew Gidden <matthew.gidden@gmail.com>

# Please confirm that this PR has done the following:

- [ ] Tests Added
- [ ] Documentation Added
- [ ] ~Name of contributors Added to AUTHORS.rst~
- [ ] ~Description in RELEASE_NOTES.md Added~

# Description of PR

Add a fast-path to `format_data` for initialization with a multi-index based Series or DataFrame that has all the required columns.

~I set the base branch for this PR to PR #730 to highlight the small additional changes necessary.~

### CODE REVIEW

File: pyam/utils.py [hunk 1]
```diff
@@ -252,41 +252,47 @@ def _format_from_database(df):
 def _intuit_column_groups(df, index):
     """Check and categorise columns in dataframe"""
 
+    existing_cols = pd.Index(df.index.names).dropna()  # skip unnamed columns
```
Comment (danielhuppmann) [inline pyam/utils.py:255]: What's the use case for unnamed columns? My intuition would be that this is invalid and should raise an error, or not?

**Comment by gidden** [comment]
Hey @coroa - I hope I didn't clobber your commits here by merging your previous PR first. Could you rebase this one on main and then I can provide a review? Thanks!

File: pyam/utils.py (hunk 1 above)
Comment (coroa) [reply pyam/utils.py:255]: Well, yes that could be better handled.

I added it during bugfix for the non-fast path case. Since the `reset_index` happens before calling to `_intuit_column_groups`, there will be an unnamed `RangeIndex`, ie. `df.index.names = [None]`.

In that case we just want to start with an empty index.

We could replace this by
```
existing_cols = pd.Index([]) if df.index.names == [None] else pd.Index(df.index.names)

**Review by gidden — APPROVED** [review]
Thanks for this @coroa - I might have a small preference on style (will add suggestion), but not blocking.

Is it possible to add a test for the untested code path?

File: pyam/utils.py [hunk 2]
```diff
@@ -344,29 +350,51 @@ def _format_data_to_series(df, index):
...
+        time_col, extra_cols, data_cols = _intuit_column_groups(df, index=index)
+
+        if isinstance(df, pd.DataFrame):
+            extra_cols_not_in_index = [c for c in extra_cols if c in df.columns]
+            if extra_cols_not_in_index:
+                df = df.set_index(extra_cols_not_in_index, append=True)
 
-    df = _convert_r_columns(df)
+            if data_cols:
+                df = df[data_cols]
```
Comment (gidden) [inline pyam/utils.py:363]: slight preference for 
```suggestion
                df = (
                    df[data_cols]
                    .rename_axis(columns=time_col)
                    .stack()
                    .rename("value")
                    )
```

Comment (coroa) [reply pyam/utils.py:363]: Sure, sure, but I'll not commit your suggestion online, since stickler won't like the hanging brace.

File: pyam/utils.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/utils.py:255]: Ok, let's leave as is and I'll try to write a few more tests for corner cases...

**Review by danielhuppmann — APPROVED** [review]
Looks good to me! If I understand it correctly, the way to access the fast-pass is to set the index before initialization, right?

File: pyam/utils.py (hunk 1 above)
Comment (coroa) [reply pyam/utils.py]: I was a bit unhappy with this and after some more thinking, i came to the conclusion it is easier to make this choice explicit by adding a `include_index` argument to `_intuit_column_groups`.

That's how it is now.

**Comment by coroa** [comment]
> Looks good to me! If I understand it correctly, the way to access the fast-pass is to set the index before initialization, right?

Indeed. (or even better keep it as it is, ie never reset it :))

**Comment by danielhuppmann** [comment]
Sorry, my earlier comment was badly phrased... What I meant was the following:

In #726, @gidden added an option `fast=False` to the IamDataFrame initialization to explicitly instruct pyam to use the fast-pass (skip some validations) - now, this is implicit. Which means the fast-pass will automatically be applied by any method using `_finalize(append=False)` (see [here](https://github.com/IAMconsortium/pyam/blob/9aa84d499900ab0d47da0a9e993a302e71483929/pyam/core.py#L225)) including aggregation and algebraic operations - but it is not possible to use the fast-pass when initializing from a file (because pandas reads a dataframe).

I think that this is perfectly fine behavior - just wanted to highlight this (or stand corrected if I'm on the wrong track).

Fine to merge (and maybe add a "force-fast-pass"-arg later). Thanks!

**Comment by coroa** [comment]
> Sorry, my earlier comment was badly phrased... What I meant was the following:
> 
> In #726, @gidden added an option `fast=False` to the IamDataFrame initialization to explicitly instruct pyam to use the fast-pass (skip some validations) - now, this is implicit. Which means the fast-pass will automatically be applied by any method using `_finalize(append=False)` (see [here](https://github.com/IAMconsortium/pyam/blob/9aa84d499900ab0d47da0a9e993a302e71483929/pyam/core.py#L225)) including aggregation and algebraic operations - but it is not possible to use the fast-pass when initializing from a file (because pandas reads a dataframe).
> 
> I think that this is perfectly fine behavior - just wanted to highlight this (or stand corrected if I'm on the wrong track).
> 
> Fine to merge (and maybe add a "force-fast-pass"-arg later). Thanks!

You are spot on. The fast-path is not improving file read-in speed (as-is), but only data passing within pyam and pandas, where the index is preserved, like with the `__finalize__` calls you are highlighting.
